### PR TITLE
Feature : Interactive shell.

### PR DESCRIPTION
**Pull request by louib**
# KeePassXC interactive shell

This PR adds an interactive shell, with Readline support, to KeePassXC.

Set of commands from the top-level executable:
```
$ ./src/cli/keepassxc-cli
Usage: ./src/cli/keepassxc-cli [options] command
KeePassXC command line interface.

Available commands:
add            Add an entry to the database.
clip           Copy an entry's password to the clipboard.
create         Create a new database.
edit           Edit a field from a database entry.
entropy-meter  Calculate password entropy.
extract        Extract and print the content of a database.
locate         Find entries quickly.
ls             List database entries.
merge          Merge two databases.
mkdir          Add a group to the database.
mv             Move an entry or a group in the database.
rm             Remove an entry from the database.
rmdir          Remove a group from the database.
shell          Launch the interactive shell.
show           Show an entry's information.


Options:
  -h, --help     Displays this help.
  -v, --version  Displays version information.

Arguments:
  command        Name of the command to execute.
```

Set of commands from the interactive shell:
```
KeePassXC 2.2.0 interactive shell
Using database My awesome passwords
Use 'help' to list the available commands.
KeePassXC> help 
add            Add an entry to the database.
clip           Copy an entry's password to the clipboard.
edit           Edit a field from a database entry.
locate         Find entries quickly.
ls             List database entries.
mkdir          Add a group to the database.
mv             Move an entry or a group in the database.
rm             Remove an entry from the database.
rmdir          Remove a group from the database.
show           Show an entry's information.
```

So basically, every shell command is also available to the top-level executable, given that the database path is provided.

I also added a `create` command so that users can create a new DB from the command line.

Also created a `Command` class, allowing for a factory-like function, which places the knowledge of all available commands at one place only (using getCommands).

## Motivation and context
I'd like to manage my passwords from the command line, and I don't think having to type in the password for every granular operation is the way to go. From what I saw with the 2.2.0 release, there seems to be an interest for a full-fledged CLI tool, so this PR is in part to see if people are interested in the feature. A shell also allows for auto-completion when searching for entries of group.

## How has this been tested?
* Manually
* Added unit tests for all the new functions in `core`.

## Questions
Some things I'm not too sure of:
* Should we build the shell as a separate executable also? I was thinking about `kpxcli`
* Any other commands you think would be useful?
* Another feedback we got from the 2.2.0 release is that the documentation for the project is lacking. With this new feature, I'd like to add documentation for the CLI tools we have. What's the plan for us to document the project? The Wiki would be nice, but it's impossible to open PRs on the wiki :(
* Right now I enable Readline support using a CMake flag. Maybe trying to find readline and enabling it if found would be better, but I'm not sure this is something CMake can do. Any suggestion is welcomed.

## Types of changes
- ✅ New feature
- ✅ Breaking change

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- X  My change requires a change to the documentation and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

### CODE REVIEW

**Comment by TheZ3ro** [comment]
>     Should we build the shell as a separate executable also? I was thinking about kpxcli.

I would prefer maintaining the List and Show command. Interactive shell are good but aren't really script friendly (compared to command-argument).

I think it's better doing something like:
`keepassxc-cli` -> Opens an interactive shell
`keepassxc-cli ./passwords.kdbx` -> Check if the first argument is a known-command, if not opens an interactive shell with `passwords.kdbx`
`keepassxc-cli show ./passwords.kdbx entry/path` -> Execute `show` command on `passwords.kdbx` to get `entry/path`

>  I'm pretty noob with CMake, so I took a file I found online and modified it for FindReadline. Is the name of the author sufficient as a copyright notice?

Needs to be added to COPYING file, I will do/review that

>  I'm having a hard time dealing with groups and entries with spaces from the shell, because of the arguments parsing based on spaces. I was thinking about converting dynamically spaces to underscores when using the shell.

I don't really like this, what's the problem with using `"` ?

>  Any other commands you think would be useful?

I don't know, we should check out KeePass CLI and see what is missing
Use #681 for that

>  Another feedback we got from the 2.2.0 release is that the documentation for the project is lacking. With this new feature, I'd like to add documentation for the CLI tools we have. What's the plan for us to document the project? The Wiki would be nice, but it's impossible to open PRs on the wiki :(

Yup, I was thinking about documentation as well. 
We can open a new repository called `wiki` and then we can make a script that pushes `wiki/master` over `https://github.com/keepassxreboot/keepassxc.wiki.git` for every new commit in `wiki/master` with a hook like @phoerious did for the website

>   Right now I enable Readline support using a CMake flag. Maybe trying to find readline and enabling it if found would be better, but I'm not sure this is something CMake can do. Any suggestion is welcomed.

IDK

>  Right now a CTRL-C would cause the shell to exit, even with changes. Intercepting the signal would be one solution, but maybe defaulting to autosave could be a better idea. It is not implemented yet.

I think the autosave (save-on-edit) will be mandatory for a thing like this. We will need to take care about conflicts like keepassxc core

**Comment by magkopian** [comment]
I would personally prefer a dedicated `-i` flag for opening an interactive shell. Usually when I have never used a command before, first thing I do is type it to the terminal without any arguments and expect to get back some basic usage information. Of course, you could argue that `-h` and `--help` flags also exist, but I just find it a bit more convenient to not having to use them.

**Comment by louib** [comment]
@TheZ3ro My goal was to make all the interactive shell commands available dynamically to the top-level executable, so that they can be used from both places.

**Comment by TheZ3ro** [comment]
@louib it's ok if we still maintain the command-argument behavior (like `keepassxc-cli list ./p.kdbx`)

**Comment by louib** [comment]
@TheZ3ro added a `Command` class. This class is used both directly by the cli and by the `Shell` command, so that new commands can be added both to the top level cli and to the shell command. This allows for reusing code between both entry points.

Creating a `Command` class also allows for a factory-like function, which places the knowledge of all available commands at one place only (using `getCommands`).

This is now the set of commands available from the top level.
```$ ./src/cli/keepassxc-cli
Usage: ./src/cli/keepassxc-cli [options] command
KeePassXC command line interface.

Available commands:
add            Add an entry to the database.
clip           Copy an entry's password to the clipboard.
create         Create a new database.
edit           Edit a field from a database entry.
entropy-meter  Calculate password entropy.
extract        Extract and print the content of a database.
ls             List database entries.
merge          Merge two databases.
mkdir          Add a group to the database.
mv             Move an entry or a group in the database.
rm             Remove an entry from the database.
rmdir          Remove a group from the database.
shell          Launch the interactive shell.
show           Show an entry's information.


Options:
  -h, --help     Displays this help.
  -v, --version  Displays version information.

Arguments:
  command        Name of the command to execute.
```

File: src/cli/AddGroup.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+#include "core/Database.h"
+#include "core/Group.h"
+#include "cli/Utils.h"
+
+AddGroup::AddGroup()
+{
+    this->name = QString("mkdir");
+    this->shellUsage = QString("mkdir new_group_path");
+    this->description = QString("Add a group to the database.");
```
Comment (weslly) [inline src/cli/AddGroup.cpp]: Can these cli strings be translatable?

Comment (louib) [inline src/cli/AddGroup.cpp]: @weslly I guess they could, yes

**Comment by louib** [comment]
@magkopian I added the `locate` command to the interactive shell and the top-level executable
```
KeePassXC> locate entry
/entry2
/entry5
/entry99
/Group2/Entry1
/trash/entry7
/trash/entry
/trash/entry"lol
/trash/entry77
/trash/entry1
```

**Pull request closed without merging**